Separate svg and gcode code from FormsUI

RafaelEstevamReis opened the pull request:
I am in progress to make LaserGRBL functions accessible in linux and via command line
The main issue is that there are a lot of dependency in a lot of classes, is hard to isolate and test
To be able to do so I moved all non-UI stuff to two Library Projects and solved most of dependency problems trying to keep the changes minimal
I made all changes in small commits to be easier to follow

In summary I:
* Separated Cyotek files from RasterConverter/Dithering to a new Lib Project
* Separated non Forms files to a new Lib Project
  * CSV folder
  * CsPotrace folder
  * SvgConverter folder
  * SvgLibrary folder
  * GrblFile
  * GrblCommand
  * Autotrace folder

I understand if this PR is too big to be revised, let me know if I can help in the process as the new library can be improved and tested more easily without any changes in the UI

arkypita commented:
Hi @RafaelEstevamReis 
Thank you for your contribution

You've done a lot of work, I don't know if I'll be able to revise and integrate it. Will I have to go on trust?

Anyway lasergrbl really needs to be rewritten. Its development thus came without planning, taking pieces and writing parts without much attention regarding maintainability and extensibility.


I would like to rewrite it by organizing everything in order to be more virtual and allow:
- a core that give all services
- different communication protocols with different firmware besides grbl (marlin but not only ...)
- virtualization / configurability of g-code writing, to allow different ways of producing on / off commands, power etc to allow use also with pen-plotter etc
- the management of different file types through "plug-ins" or modules that can be added simply by implementing a class / interface without necessarily having to deal with the entire project
- same modular approach for different raster processing tools, so you can add other options to "line2line, dithering, vectorization, centerline"
- saving of configuration presets
- manage job as project where you can add different part, set different process for each part, save as project instead of g-code so you can open and edit later etc

yes... and many more!

This is much more than a simple refactoring, It requires a real rewrite and currently I don't have the time to be able to embark on this job.

About your work I really appreciate the initiative to do a little order. If the ultimate goal is to use it in linux I wish you to be able to find a solution to the problems with the UI.

I had tried mono / sharpdevelop but had too many problems. Maybe today things have improved, however it is good to have a well organized project. I promise that when I can I will take a look at your commit

Patola commented:
@arkypita 
please accept Rafael's contributions. I am following closely with him. I just recently got a 3D printer that can also do laser engraving (BIBO 2 Wifi) and I only use Linux, so I am testing the build Rafael sends me. He's a long time friend.

RafaelEstevamReis commented:
Most of the "file count changed" was separating the "LaserGRBL contains some code from:" part from Readme to different projects (simple moves)

Most of changed files was separating files with multiple classes to one class per file (I had to do this to better understand the relations)
I tried to don't change anything algorithm-wise, but I admit that without any unit-tests this is a hard claim

I my branch I:
* separated even more
* Migrated everything to DotNet-Core
* Created a Command-Line Console-UI (as today does almos nothing)

I booted a Linux Vm and could compile all libs an the Console-UI without mono neither wine, using the microsoft's linux dot-net core version in Visual Studio Code
Microsoft's code doesn't have linux support for Forms, but the rest being ported and running makes the goal closer

The most critical change I made to port for Core was changing the GCodeFromSVG.cs, this class uses the System.Windows.Media namespace which is not available. But it uses only to access Point and Matrix, I could use System.Drawing.2D.PointF to replace Point and ported (copied with some changes) the Mono version of the  System.Windows.Media.Matrix the apply the transforms

One thing I could not port was the MS-SVG library, it depends on much more stuff. As it is used ONLY to call Autotrace and only for CenterLine mode I give-up. Must be be better library now days to do that

Onde (somewhat) dangerous thing I have to change was Settings reading GrblCore static fields to get settings' location
I migrated GrblFile and GrblCommand to the library (UI independant) but GrblCore is too entangled and I don't understand it enough to touch

RafaelEstevamReis commented:
One thing that I have to change but don't understand enough to touch was: There are a LOT of P/invoke (kernel, user32, etc) calls 

I found two places where it is heavily used, the HiResTimer witch I do not touch for fear of breaking the GCode stream process
GrblFile used the timer to measure load/save times, I replace to a relatively "lowres" (timespan by DateTime.Now subtraction) and make it independent from the Calls
On linux, as you suggested on some thread I replaced with a StopWatch and faked some frequency measuraments. I have to admit that the Italian or French comments on this class made me uncomfortable to change anything (my primary language is portuguese) and I do not trust google translate for a such complex class

The other heavy of of P/Invoke was  on OS Fingerprinting which seemed only used on Bug Report
I Changed some calls to Enviroment.[Stuff] and do not touch the rest

My last commit on my branch allow-me to convert via command line SVG to GCode and bitmap to GCode on l2l mode. I do not implement the settings because is too tied to the UI, ideally the UI should separate it first

As the UI load the config to ImageProcessor and then fill the component's value, I think is possible to ImageProcessor.cs load/save the config and UI read/save to ImageProcessor (or even use a separate class to expose this)

RafaelEstevamReis commented:
Finally...
On safety-trust matter, I think is possible to make some integration (end-to-end) tests e compare the results before and after the merge to check if identical gcode are being generated. Is not bullet proof, but can be a good start point (an easier one with command line capabilities).
